### ISSUE TITLE
minecraftbe.service Failed after daily restart

### ISSUE THREAD

**Issue by magicmatteo** [TheRemote/MinecraftBedrockServer#98]
I am monitoring the state of the minecraft server with something called Zabbix. This gives me the ability to monitor the status of a systemd service and send alerts if it goes down etc.

After the daily restart - the service is reported as failed.

A possible fix:
Instead of running start.sh at the end of restart script, run systemctl start server.service, which calls start.sh anyway.
Would require allowing non root user to start service, which gets messy.

You might have some better ideas on how to make the service work correctly.

**Comment by TheRemote** [comment]
Hey magicmatteo,

That's definitely a tricky one, and I definitely can see what you're talking about in the code and I always tell people they need to start it with systemd if they track the server by service (like Zabbix is).

It's considered inappropriate to try to rig a forking process to work in this way (not being launched with systemd and systemd somehow figuring it out and bringing the service online) with systemd.  Most people will say it's impossible and those who don't say you shouldn't ever do it and you need to redesign your service with a more appropriate service type/model.

That would probably leave the best option of allowing the server to be started with a "passwordless" sudo by adding this line to sudoers with

`sudo visudo`

and inserting:

`minecraftuser ALL=(ALL) NOPASSWD: /bin/systemctl start yourservername`

At this point you could simply swap out the start.sh line in restart.sh with the line you posted before.

Automating this is messy as you stated.  Sudoers is designed to be protected from automated changes like this and while not undefeatable I'm wondering if this is something that could be added to the documentation, or the script could even give them the line to add to the sudoers file, there's a couple different approaches.

Would you be willing to try the change I outlined and see if it solves it for you?  I will definitely think some more about how to approach this as it's not unreasonable for the restart.sh script to function with the service correctly in "online" mode but it's a tricky problem to solve for sure.  Thanks for reporting this and let me know if that fix works appropriately with your setup!

**Comment by magicmatteo** [comment]
Thanks for the quick response! Yes, after lodging this issue, I got to work on coming up with something a bit more constructive than just outlining my problem.

I ended up doing exactly what you've suggested.

Added my user account to sudoers using `sudo visudo` - although people suggest not editing the main sudoers file and instead adding your edits to its own file in /etc/sudoers.d/_myserver.conf_. This will provide as you suggested, passwordless sudo capabilities for each particular command you list in here. Note that I did not do this though - I just used sudo visudo.

The following command inside sudoers worked for me:
`minecraftuser ALL= NOPASSWD: /bin/systemctl start yourservername`

I then replaced the final line (67) in restart.sh from:
`/bin/bash dirname/minecraftbe/servername/start.sh`
To:
`sudo systemctl start yourservername.service`

Can confirm its currently working as intended, did a fair bit of testing with it last night. Calling restart.sh brings the service down and then back up correctly. Currently have Zabbix triggering an email alert if the service is down for more than 3 minutes.

As you say, it's tricky to automate based on the sensitivity of systemd. My linux knowledge certainly goes no further than just controlling services with systemctl. I've got no idea of the inner workings of it all. I wish I could help a bit more.

**Comment by TheRemote** [comment]
Outstanding!  So for now I've added it to the documentation, as well as put a precommented out line with the instructions to enable it using the sudo line.  I figured this was the safest option for now without going in and automatically adding sensitive sudo entries but I'm open to automating it using the /etc/sudoers.d/ config folder in the future here potentially as that's quite a bit less invasive, doesn't touch any other config files and is also easy to remove.

Thanks for following up on this!  The changes are pretty much pre-done for you in the scripts (for the future like potential updates/fixes) which should make it much easier to know what to do to enable this functionality and if this continues to come up then I'll weigh it as worth automating for sure.  Thanks again!
